### ISSUE TITLE
Confusion regarding YOLO region layer replacement

### ISSUE THREAD

**Issue by IAI-ilyasem** [openvinotoolkit/openvino#4546]
Hello,

I wanted to ask about how the YOLO Region layer replacement works with regard to original Darknet parameters, specifically with respect to the sigmoid function and scale_x_y parameter.

In this: https://docs.openvinotoolkit.org/latest/omz_models_public_yolo_v3_tf_yolo_v3_tf.html
It is said the OpenVINO converted model specifically takes care of needing to use the sigmoid function to find bounding box x,y (applied to tx,ty relative to anchor). How is this done? It is not clear in the following files:

openvino/model-optimizer/extensions/front/YOLO.py
openvino/model-optimizer/extensions/ops/regionyolo.py

In this: https://github.com/AlexeyAB/darknet/issues/5168
AlexeyAB confirms that the scale_x_y parameter (found in the darknet .cfg file, not found in supported_attrs or backend_attrs of regionyolo.py) has the following effect on bounding box x,y:

bx[i] = (sigmoid(tx[i]) - 0.5) * scale_x_y + 0.5 + c[i]

The equation found in  object_detection_demo_yolov3_async.py however is (converted to same format):

bx[i] = tc[i] + c[i]

If the darknet model is created with (scale_x_y != 1), how should this accounted for in the conversion to IR?

Thank you for considering my question!

**Comment by Iffa-Meah** [comment]
Hi,

Please take note that the object_detection_demo_yolov3_async.py is already deprecated in the latest release.
This is the new demo for [Object Detection](https://github.com/openvinotoolkit/open_model_zoo/blob/master/demos/python_demos/object_detection_demo/object_detection_demo.py)

You could download the model by using the command:
python "C:\Program Files (x86)\Intel\openvino_2021\deployment_tools\tools\model_downloader\downloader.py" --name yolo-v3-tf

and convert it:
python "C:\Program Files (x86)\Intel\openvino_2021\deployment_tools\tools\model_downloader\converter.py" --name yolo-v3-tf
Note: This is so far tested in Windows, it might differ in Linux

If you take a look at the commands that came after running the converter.py, you will notice that yolo-v3.json is used in the transformation_config. This means, during model conversion using the MO, the TF model has been converted with the .json configuration downloaded with the model downloader/converter.

The command also should address the scale_values parameter

Do you have any issues with the convertion or inferencing?


Sincerely,
Iffa

**Comment by IAI-ilyasem** [comment]
Hi Iffa,

Thanks for considering my question.

I have custom trained darknet yolov4 models with some variation in the scale_x_y parameter. I am able to convert and run inference normally, but am seeing some inconsistency related to that term.


This new demo code seems to have a very similar function for parsing the output embedded in [yolo.py](https://github.com/openvinotoolkit/open_model_zoo/blob/338630987b403a6981d03ab6d04c2d5ad367793a/demos/python_demos/common/models/yolo.py#L23) as staticmethod _parse_yolo_region()

```
            x = (col + x) / params.sides[1]
            y = (row + y) / params.sides[0]
```

My confusion comes from the fact that the transformation config .json file passed to the model optimizer during conversion does not seem to support something related to scale_x_y in "custom attributes". I am looking at [regionyolo.py](https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/ops/regionyolo.py) and referencing:

```
    def supported_attrs(self):
        return [
            'coords',
            'classes',
            'num',
            'axis',
            'end_axis',
            'do_softmax',
            'anchors',
            'mask'
        ]
```

This is related as class Params in  [yolo.py](https://github.com/openvinotoolkit/open_model_zoo/blob/338630987b403a6981d03ab6d04c2d5ad367793a/demos/python_demos/common/models/yolo.py#L23) sees the same values for coords, classes, num, anchors, and mask as is passed in the json.

**Does the operation support a parameter for scale_x_y? If so what is the full list of supported "custom attributes" for TFYOLOV3?**


Are you referring to the scale_values parameter that can be used with the mo.py command? See here: [cli_parser.py](https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/mo/utils/cli_parser.py)

```
    common_group.add_argument('--scale_values',
                              help='Scale values to be used for the input image per channel. ' +
                                   'Values are provided in the (R,G,B) or [R,G,B] format. ' +
                                   'Can be defined for desired input of the model, for example: ' +
                                   '"--scale_values data[255,255,255],info[255,255,255]". ' +
                                   'The exact meaning and order ' +
                                   'of channels depend on how the original model was trained.',
                              default=())
```

This doesn't seem to match the description of scale_x_y in darknet.


Best,

-Ilya

**Comment by Iffa-Meah** [comment]
Generally, If you look at the main code of the model optimizer, the way it perform is, it would automatically adjust the scale according to the model it had directed to (this should be related to the YOLO Neural Network too) and the model optimizer uses all the info to transform them into IR.

Btw, When a network is defined in Python* code, you have to create an inference graph file. Usually graphs are built in a form that allows model training. That means that all trainable parameters are represented as variables in the graph. To be able to use such graph with Model Optimizer ,such graph should be frozen. This is where you need to use the coco and weight file to produce frozen model (.pb). This is how all traits of the original model are inherited by the IR.

Hence, my opinion is, scale_x_y in darknet depends on the model's Neural Network since a frozen model would be generated based on it. From OpenVINO side, it just fetch the traits and transform them into IR. 

One of the way to define your own scale is by using the parameter --scale_values as I mentioned before (to change the scale values for the input image per channel). This values are actually provided in the (R,G,B) or [R,G,B] format.
[This](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Converting_Model_General.html) might be helpful to you.

You also need to be sure your yolo is in which category, if you are using yolo from darknet and convert it to IR you may use the step I provided before OR [this](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_YOLO_From_Tensorflow.html).

If you are trying to convert keras yolo to IR,  you will need to convert from Keras to frozen TF then use that frozen model to convert to IE IR using model optimizer.

**Comment by Iffa-Meah** [comment]
Closing issue, feel free to re-open or start a new issue if additional assistance is needed
